### PR TITLE
Mypy strictness

### PR DESCRIPTION
**Pull request by hukkin**
- Add mypy strictness flags:
    ```
    warn_unreachable = true
    warn_unused_ignores = true
    warn_redundant_casts = true
    ```
- Fix issues found by mypy

### CODE REVIEW

File: src/marshmallow/schema.py [hunk 1]
```diff
@@ -1025,9 +1025,13 @@ def _bind_field(self, field_name: str, field_obj: ma_fields.Field) -> None:
...
         except TypeError as error:
-            # field declared as a class, not an instance
-            if isinstance(field_obj, type) and issubclass(field_obj, base.FieldABC):
-                msg = (
+            # Field declared as a class, not an instance. Type ignore because
+            # we handle unsupported arg types -> this is dead code from
+            # typechecker's perspective.
+            if isinstance(field_obj, type) and issubclass(  # type: ignore
+                field_obj, base.FieldABC
+            ):
```
Comment (hukkin) [inline src/marshmallow/schema.py:1033]: The formatting here is awkward, but the only formatting that passes both flake8 line length and black checks. Another option is to keep the `if` as a one-liner and add `# noqa: E501` after the `# type: ignore` to disable flake8 line length check.

Comment (hukkin) [reply src/marshmallow/schema.py:1033]: Also, I don't know what the importance of the custom error message here is, but perhaps we could consider not handling this TypeError at all and just let the original exception bubble up always?

Comment (sloria) [reply src/marshmallow/schema.py:1033]: It makes debugging a bit easier. The original exception message doesn't really pinpoint what the issue is.

**Review by sloria — APPROVED** [review]
Thank you!

**Pull request merged**